src/core/thread/router_table.cpp

```diff
@@ -0,0 +1,629 @@
+#include "thread/thread_netif.hpp"
+
+namespace ot {
+
+RouterTable::Iterator::Iterator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mRouter(NULL)
+{
+    Reset();
+}
+
+void RouterTable::Iterator::Reset(void)
+{
+    RouterTable &routerTable = GetInstance().Get<RouterTable>();
+
+    mRouter = &routerTable.mRouters[0];
+}
+
+void RouterTable::Iterator::Advance(void)
+{
+    RouterTable &routerTable = GetInstance().Get<RouterTable>();
+    Router *     listEnd     = &routerTable.mRouters[Mle::kMaxRouters];
+
+    VerifyOrExit(mRouter != NULL);
+
+    mRouter++;
+
+    VerifyOrExit(mRouter < listEnd && mRouter->GetRloc16() != 0xffff, mRouter = NULL);
+
+exit:
+    return;
+}
+
+RouterTable::RouterTable(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mRouterIdSequenceLastUpdated(0)
+    , mRouterIdSequence(Random::GetUint8())
+{
+    Clear();
+}
+
+void RouterTable::Clear(void)
+{
+    memset(mAllocatedRouterIds, 0, sizeof(mAllocatedRouterIds));
+    memset(mRouterIdReuseDelay, 0, sizeof(mRouterIdReuseDelay));
+    UpdateAllocation();
+}
+
+void RouterTable::ClearNeighbors(void)
+{
+    for (uint8_t i = 0; i < Mle::kMaxRouters; i++)
+    {
+        mRouters[i].SetState(Neighbor::kStateInvalid);
+    }
+}
+
+bool RouterTable::IsAllocated(uint8_t aRouterId) const
+{
+    return (mAllocatedRouterIds[aRouterId / 8] & (1 << (aRouterId % 8))) != 0;
+}
+
+void RouterTable::UpdateAllocation(void)
+{
+    uint8_t indexMap[Mle::kMaxRouterId + 1];
+    uint8_t routerCount = 0;
+
+    // build index map
+    for (uint8_t i = 0; i <= Mle::kMaxRouterId; i++)
+    {
+        if (IsAllocated(i))
+        {
+            indexMap[i] = routerCount++;
+        }
+        else
+        {
+            indexMap[i] = Mle::kInvalidRouterId;
+        }
+    }
+
+    // shift entries forward
+    for (int i = Mle::kMaxRouters - 2; i >= 0; i--)
+    {
+        uint8_t routerId = mRouters[i].GetRouterId();
+        uint8_t newIndex;
+
+        if (routerId > Mle::kMaxRouterId || !IsAllocated(routerId))
+        {
+            continue;
+        }
+
+        newIndex = indexMap[routerId];
+
+        if (newIndex > i)
+        {
+            mRouters[newIndex] = mRouters[i];
+        }
+    }
+
+    // shift entries backward
+    for (uint8_t i = 1; i < Mle::kMaxRouters; i++)
+    {
+        uint8_t routerId = mRouters[i].GetRouterId();
+        uint8_t newIndex;
+
+        if (routerId > Mle::kMaxRouterId || !IsAllocated(routerId))
+        {
+            continue;
+        }
+
+        newIndex = indexMap[routerId];
+
+        if (newIndex < i)
+        {
+            mRouters[newIndex] = mRouters[i];
+        }
+    }
+
+    // fix replaced entries
+    for (uint8_t i = 0; i <= Mle::kMaxRouterId; i++)
+    {
+        uint8_t index  = indexMap[i];
+        Router &router = mRouters[index];
+
+        if (index == Mle::kInvalidRouterId)
+        {
+            continue;
+        }
+
+        if (router.GetRouterId() != i)
```

Comment:
How can we get to this situation? 
If I understand the code correctly, it's handling a router id which is in the allocated list and has a valid `indexMap` but  its corresponding `Router` entry does not have the same router id.  Is it only applicable for the router id used by the device itself?

Reply:
This situation occurs when a router ID has been removed and a new router ID takes its place.  This check serves to indicate whether the router entry needs to be reinitialized.

For example:
Old Router IDs: {1, 2, 4}
New Router IDs: {1, 3, 4}

We need to reinitialize `mRouters[1]` because router ID 2 is now longer valid and router ID 3 is a brand new router that was just introduced.